src/sumcheck/sumcheck_small_value.rs

```diff
@@ -0,0 +1,841 @@
+        })
+        .reduce(|| Accumulators::<EF>::new_empty(), |a, b| a + b) 
+}
+
+// Given a point w = (w_1, ..., w_l), it returns the evaluations of eq(w, x) for all x in {0, 1}^l.
+pub fn eval_eq_in_hypercube<F: Field>(point: &Vec<F>) -> Vec<F> {
+    let n = point.len();
+    let mut out = F::zero_vec(1 << n); 
+    eval_eq_batch::<F, F, false>(
+        RowMajorMatrixView::new_col(&point),
+        &mut out,
+        &[F::ONE],
+    );
+    out 
+}
+
+// Given two multilinear points p and q, it evaluates eq(p, q).
+pub fn eval_eq_in_point<F: Field>(p: &[F], q: &[F]) -> F {
+    let mut acc = F::ONE;
+    for (&l, &r) in p.into_iter().zip(q) {
+        acc *= F::ONE + l * r.double() - l - r;
+    }
+    acc
+}
+
+// Given the points w and r, we compute the linear function 
+// l(x) = eq( w_1,...w_{i-1} ; r_1,...r_{i-1} ) * eq(w_i, x)
+// Section 5.1, page 17.
+pub fn compute_linear_function<F: Field>(w: &[F], r: &[F]) -> [F; 2] {
+    let round = w.len();
+    debug_assert!(r.len() == round - 1);
+
+    let mut const_eq: F = F::ONE;
+    if round != 1 {
+        const_eq = eval_eq_in_point(&w[..round - 1], r);
+    }
+    let w_i = w.last().unwrap();
+
+    // Evaluation of eq(w,X) in [eq(w,0),eq(w,1)]
+    [const_eq * (F::ONE - *w_i), const_eq * *w_i]
+}
+
+// Given the linear functions l and t, we compute S(0) and S(inf).
+// See Procedure 8, Page 35.
+fn get_evals_from_l_and_t<F: Field>(l: &[F; 2], t: &[F]) -> [F; 2] {
+    [
+        t[0] * l[0],                   // S(0)
+        (t[1] - t[0]) * (l[1] - l[0]), // S(inf) -> l(inf) = l(1) - l(0)
+    ]
+}
+
+// Algorithm 6. Page 19.
+// Compute three sumcheck rounds using the small value optimizaition and split-eq accumulators.
+// It returns the three challenges r_1, r_2, r_3.
+pub fn svo_three_rounds<Challenger, F: Field, EF: ExtensionField<F>>(
+    prover_state: &mut ProverState<F, EF, Challenger>,
+    poly: &EvaluationsList<F>,
+    w: &MultilinearPoint<EF>,
+    sum: &mut EF,
+) -> (EF, EF, EF)
+where
+    Challenger: FieldChallenger<F> + GrindingChallenger<Witness = F>,
+{
+    let e_in = precompute_e_in(w);
+    let e_out = precompute_e_out(w);
+
+    // We compute all the accumulators A_i(v, u).
+    let accumulators = compute_accumulators(poly, e_in, e_out);
+
+    // ------------------   Round 1   ------------------
+
+    // 1. For u in {0, 1} compute t_1(u)
+    // Recall: In round 1, t_1(u) = A_1(u).
+    let t_1_evals = accumulators.get_accumulators_for_round(0);
+
+    // 2. For u in {0, 1, inf} compute S_1(u) = t_1(u) * l_1(u).
+
+    // We compute l_1(0) and l_1(1)
+    let linear_1_evals = compute_linear_function(&w.0[..1], &[]);
+
+    // We compute S_1(0) and S_1(inf)
+    let round_poly_evals = get_evals_from_l_and_t(&linear_1_evals, &t_1_evals);
+
+    // 3. Send S_1(u) to the verifier.
+    prover_state.add_extension_scalars(&round_poly_evals);
+
+    // 4. Receive the challenge r_1 from the verifier.
+    let r_1: EF = prover_state.sample();
```

Comment:
We should apply grinding here as well no?

```suggestion
    prover_state.pow_grinding(pow_bits);
    let r_1: EF = prover_state.sample();
```